neurodamus/metype.py

```diff
@@ -10,6 +10,86 @@
 from .core.configuration import ConfigurationError, SimConfig
 
 
+class SectionIdError(Exception):
+    pass
+
+
+_section_layout = [
+    ("soma", lambda c: c.soma, lambda c: int(c.nSecSoma)),
+    ("axon", lambda c: c.axon, lambda c: int(c.nSecAxonalOrig)),
+    ("dend", lambda c: c.dend, lambda c: int(c.nSecBasal)),
+    ("apic", lambda c: c.apic, lambda c: int(c.nSecApical)),
+    ("ais", lambda c: getattr(c, "ais", []), lambda c: int(getattr(c, "nSecLastAIS", 0))),
+    ("node", lambda c: getattr(c, "node", []), lambda c: int(getattr(c, "nSecNodal", 0))),
+    ("myelin", lambda c: getattr(c, "myelin", []), lambda c: int(getattr(c, "nSecMyelinated", 0))),
+]
+
+
+def get_section_id(cell, section):
+    """Calculate the global index of a given section within its cell.
+
+    :param cell: The cell instance containing the section of interest
+    :param section: The specific section for which the index is required
+    :return: The global index of the section, applicable for neuron mapping
+
+    Note: section_id is based on the original cell, before removing the axon.
+
+    Warning: this method returns the original section_id, which may not be valid
+    if the axon was removed. The offsets are still calculated based on the
+    original cell structure.
+    """
+    section_name = str(section).rsplit(".", 1)[-1]
+    try:
+        section_type, index_str = section_name.rsplit("[", maxsplit=1)
+        local_idx = int(index_str.rstrip("]"))
+        if local_idx < 0:
+            raise SectionIdError(f"Negative index {local_idx} in section name: {section_name}")
+    except ValueError as e:
+        raise SectionIdError(f"Cannot parse section name: {section_name}") from e
+
+    offset = 0
+    for name, _, count_fn in _section_layout:
+        count = count_fn(cell)
+        if name == section_type:
+            if local_idx >= count:
+                raise SectionIdError(
+                    f"Index {local_idx} out of range for section type '{name}' (count={count})"
+                )
```

Comment:
This check will make many current simulations fail because @darshanmandge mentioned that they won't update all the emodel files in "Zenodo, other public repositories for different circuits"
Can we apply a workround here? if `local_idx >= count(which is c.nSec<section_type>)` for a certain `section_type`, we count the length of `c.<section_type>` to get the correct `count`.
Another (better) solution would be that we count the length of every `c.<section_type>` , meaning we don't use c.nSec<section_type>.  For that, I think we can count per emodel+morph_name combination. Cells instantiated with the same emodel+morph_name share the same section counts.

Reply:
-About the 2nd solution, why are you against counting the section in neurodamus? You did that in your previous commit in this PR. I thought we agreed not to use the nSecxxx from the emodel hoc templates. Ok, we can discuss again.

I understood the exact opposite from our previous conversation